### PR TITLE
EVAKA-4190 Thread replies

### PR DESCRIPTION
**Pull request by kimifoud**
#### Summary

- Implement thread reply functionality in the backend
- Prevent users from seeing messages where they are not a participant
- Deny replies to bulletins

### CODE REVIEW

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import fi.espoo.evaka.shared.db.Database
+import fi.espoo.evaka.shared.domain.Forbidden
+import fi.espoo.evaka.shared.domain.NotFound
+import org.springframework.stereotype.Service
+import java.util.UUID
+
+@Service
+class MessageService {
+    fun replyToThread(
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageService.kt:15]: This function could probably be just a top level function in `Message.kt`. At least right now I can't think of any benefit of making it a service unless maybe if you want to mock it or use dependency injection.

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/Message.kt [hunk 2]
```diff
@@ -54,13 +55,20 @@ fun createMessageThread(
     content: String,
     type: MessageType,
     sender: MessageAccount,
-    recipientAccountIds: Set<UUID>
+    recipientAccountIds: Set<UUID>,
```
[the remark this answers is not in the document]
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/Message.kt:58]: I noticed this too 🤔  `./gradlew formatKotlin && ./gradlew lintKotlin` was run before the commit.

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageController.kt [hunk 3]
```diff
@@ -92,6 +96,37 @@ class MessageController {
...
+    )
+
+    @PostMapping("/{messageId}/reply")
+    fun replyToThread(
+        db: Database.Connection,
+        user: AuthenticatedUser,
+        @PathVariable messageId: UUID,
+        @RequestBody body: ReplyToMessageBody,
+    ) {
+        Audit.MessagingReplyToMessageWrite.log()
```
[the remark this answers is not in the document]
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageController.kt:112]: Good point, will do!

File: service/src/main/kotlin/fi/espoo/evaka/messaging/message/Message.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/Message.kt:60]: IDEA seems to be adding those on auto-format after version update. The linter does not seem to care about trailing commas.

Ran `./gradlew ktlintApplyToIdea` again and restarted. Now it does not add them anymore.

[the remark this answers is not in the document]
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/Message.kt:60]: However, neither `./gradlew formatKotlin` nor IDEA remove the trailing commas. Is there something wrong with the linter?

[the remark this answers is not in the document]
Comment (kimifoud) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/message/Message.kt:60]: It seems like there's no trailing comma support: https://github.com/pinterest/ktlint/issues/709

**Review by Joosakur — APPROVED** [review]
💎

**Pull request merged**